### ISSUE TITLE
components in v2 addon's app-js using index files broken

### ISSUE THREAD

**Issue by simonihmig** [embroider-build/embroider#1835]
Given index-modules based file layout for colocated component files as supported by RFC481 in a v2 addon:

```
src
└── componennts
    └── fancy-button
        ├── index.ts
        └── index.hbs
```

And generating the package's `app-js` by the rollup plugin provided by `@embroider/addon-dev`:

```json
{
  "ember-addon": {
    "version": 2,
    "type": "addon",
    "main": "addon-main.cjs",
    "app-js": {
      "./components/fancy-button/index.js": "./dist/_app_/components/fancy-button/index.js"
    }
  }
}
```

when building an app consuming the addon using Embroider v3.4.5 and "optimized" settings, it would fail with:

```
Module not found: Error: Can't resolve '#embroider_compat/components/fancy-button' in '[...]/test-app'
```

It *does* work work with either
* classic build w/ eai
* Emboider v2.x
* Embroider v3.4.5 w/ "safe" mode (no static components)

It will also work in Embroider v3.4.5 "optimized" when changing the app-js's left side to not use index modules:
```diff
-"./components/fancy-button/index.js": "./dist/_app_/components/fancy-button/index.js"
+"./components/fancy-button.js": "./dist/_app_/components/fancy-button/index.js"
``` 

So we _could_ fix this by exposing the app-js not using index modules, but since index-modules are officially supported by RFC481, it looks like a bug in Embroider?

**Comment by NullVoxPopuli** [comment]
I think this can be customized via:
https://github.com/embroider-build/embroider/blob/main/packages/addon-dev/src/rollup-app-reexports.ts#L11

```js
addon.appReexports(['**/*.js'], { mapFilename: (filename) => {
  if (filename.includes('/components/') && filename.endsWith('index.js')) {
    return path.dirname(filename) + '.js';
  }
} }),
```

does that work?

See also prior discussions:
- https://github.com/embroider-build/embroider/issues/1220
- https://github.com/embroider-build/embroider/pull/1639
  - https://github.com/embroider-build/embroider/pull/1587
  - https://github.com/embroider-build/embroider/pull/1596

tl;dr: we opted not to resolve this at the resolver layer, since it's an explicit `app-js` and package.json exports resolution configuration that causes the behavior you're seeing.

**Comment by simonihmig** [comment]
@mansona are you able to point me to where this was fixed on `main`?

**Comment by simonihmig** [comment]
> I think this can be customized via:

Yeah, I know this workaround exists. But the question in this issue was more where to _properly_ fix this? Is this a bug in Embroider (seems so, as Chris confirmed), or should we bake that behaviour into the rollup plugin? I'd rather not want to introduce this workaround everywhere in user-land code, since index-modules is a supported thing.

**Comment by NullVoxPopuli** [comment]
>  since index-modules is a supported thing

Not in ESM node :sweat_smile: (which is a bit of an oversight for co-location -- I _also_ learned this by running in to broken stuff).
I think the fix @mansona is talking about is the loose mode resolver rules, which had index behavior fixed here: https://github.com/embroider-build/embroider/pull/1794
Commented here: https://github.com/embroider-build/embroider/pull/1826#pullrequestreview-1908642381
Actual fix here: https://github.com/embroider-build/embroider/pull/1794/files#diff-61967680c203e14cb9a3b2c252caca7d6b0ee9b285991dba6b728f6c54112a12R571

**Comment by amk221** [comment]
> I'd rather not want to introduce this workaround everywhere in user-land code, since index-modules is a supported thing.

Agree, I don't want to have to manually manage this code in multiple addons. Can't ember fix this for me?

**Comment by NullVoxPopuli** [comment]
This is just how ESM javascript works -- I don't think there is anything to fix.


Here is a demo of a v2 addon using the workaround

https://stackblitz.com/edit/stackblitz-starters-apl1kjs6?description=Starter%20project%20for%20Node.js,%20a%20JavaScript%20runtime%20built%20on%20Chrome%27s%20V8%20JavaScript%20engine&file=.gitignore,my-addon%2Frollup.config.mjs&file=my-addon%2Fpackage.json,my-addon%2Fdist%2F_app_%2Fcomponents%2Ffoo.js&title=node.new%20Starter
